Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,172 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Metrics API open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the Swift Metrics API project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Metrics API project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Combine
+import Foundation
+
+@available(OSX 10.15, *)
+public enum SystemMetricsProvider {
+    fileprivate static let lock = ReadWriteLock()
+    fileprivate static let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsProvider", qos: .background)
+    fileprivate static var timeInterval: DispatchTimeInterval = .seconds(2)
+    fileprivate static var cancellable: Cancellable?
+    fileprivate static var systemMetricsType: SystemMetrics.Type = UnsuportedMetricsProvider.self
+
+    public static func bootstrapSystemMetrics(pollInterval interval: DispatchTimeInterval = .seconds(2), systemMetricsType: SystemMetrics.Type? = nil) {
+        self.lock.withWriterLockVoid {
+            self.timeInterval = interval
+            if let type = systemMetricsType {
+                self.systemMetricsType = type
+            } else {
+                #if os(Linux)
+                self.systemMetricsType = LinuxSystemMetricsProvider.self
+                #else
+                self.systemMetricsType = UnsuportedMetricsProvider.self
+                #endif
+            }
+        }
+        self.updateSystemMetrics()
+    }
+
+    public static func cancelSystemMetrics() {
+        self.cancellable?.cancel()
+    }
+
+    fileprivate static func updateSystemMetrics() {
+        let interval = self.lock.withReaderLock { self.timeInterval }
+        let c = self.queue.schedule(after: .init(.now()), interval: .init(interval)) {
+            let prefix = "process_"
+            let pid = ProcessInfo.processInfo.processIdentifier
+            let metrics = self.systemMetricsType.init(pid: "\(pid)")
+            if let vmem = metrics.virtualMemory { Gauge(label: prefix + "virtual_memory_bytes").record(vmem) }
+            if let rss = metrics.residentMemory { Gauge(label: prefix + "resident_memory_bytes").record(rss) }
+            if let start = metrics.startTimeSeconds { Gauge(label: prefix + "start_time_seconds").record(start) }
+            if let cpuSeconds = metrics.cpuSeconds { Gauge(label: prefix + "cpu_seconds_total").record(cpuSeconds) }
+            if let maxFds = metrics.maxFds { Gauge(label: prefix + "max_fds").record(maxFds) }
+            if let openFds = metrics.openFds { Gauge(label: prefix + "open_fds").record(openFds) }
+        }
+        self.lock.withWriterLockVoid {
+            self.cancellable = c
+        }
+    }
+}
+
+public protocol SystemMetrics {
+    init(pid: String)
```

Comment:
Passing PID as string seems a bit weird, pass as Int and string it inside the impl?

Reply:
If we’re doing this, we need to also think about Windows, so paging @compnerd

Reply:
Windows uses a `DWORD` (type alias for Int32).  Id prefer `process` or `id` rather than `pid` since `pid` and `pid_t` are pretty Unix centric.

I completely agree with @ktoso that `String` is an odd type for this.  Why not have something like:

```
#if os(Windows)
typealias ProcessID = DWORD
#else
typealias ProcessID = pid_t
#endif
```

Reply:
I have gone with @compnerd 's solution here, but wanted to validate that on windows ` ProcessInfo.processInfo.processIdentifier` will return a `DWORD`, otherwise we need an extra check here: https://github.com/apple/swift-metrics/pull/61/files#diff-b35e7c2396a198ceca87c72cdc27817fR43